tests/sentry/auth/test_idpmigration.py

```diff
@@ -18,5 +25,26 @@ def test_create_verification_key(self, send_confirm_email):
         assert send_confirm_email.call_args.args[1] == self.email
         assert len(send_confirm_email.call_args.args[2]) == 32
 
-    # TODO
-    # def test_verify_new_identity(self):
+    @mock.patch("sentry.auth.idpmigration.send_confirm_email")
+    def test_verify_new_identity(self, send_confirm_email):
+        idpmigration.create_verification_key(self.user, self.org, self.email)
+        path = reverse(
+            "sentry-idp-email-verification",
+            args=[send_confirm_email.call_args.args[2]],
+        )
+        response = self.client.get(path)
+        assert (
+            self.client.session["verification_key"]
+            == f"auth:one-time-key:{send_confirm_email.call_args.args[2]}"
+        )
+        assert response.status_code == 302
+        assert response.url == "/auth/login/"
+
+    def test_verify_new_identity_wrong_key(self):
+        idpmigration.create_verification_key(self.user, self.org, self.email)
+        path = reverse(
+            "sentry-idp-email-verification",
+            args=[get_random_string(32, string.ascii_letters + string.digits)],
+        )
```

Comment:
🤔 Nondeterministic behavior in unit tests generally sets off alarm bells for me, but this is probably okay. There can be trouble if
1. randomly chosen values show up in the expected/actual values of assertions, causing log messages to differ annoyingly between test runs; or
2. there is a non-astronomical probability of a spurious pass or fail if a certain value is chosen by coincidence.

Neither of these apply here (the assertion compares only status codes, and the chance of a false negative is equal to that of the security feature being broken by brute force). But, for situations where it does, it's good to think the random behavior being by mocked out.

Reply:
yeah, +1 to mocking out a "wrong key"

Reply:
Putting a constant "wrong key" is an improvement.

Not to belabor my point, but the real challenge is that `send_one_time_account_confirm_link` itself is nondeterministic because it contains `get_random_string`. So the test could theoretically have a false failure if it generates `"d14Ja9N2eQfPfVzcydS6vzcxWecZJG2z2"` as the string. But that's infeasible.

If it weren't infeasible, options would include patching over `get_random_string` in the function under test, or inspecting value of the `verification_key` returned to the test and constructing a wrong key that is known to be unequal to it.